src/chainrules.jl

```diff
@@ -1,8 +1,43 @@
-function ChainRulesCore.rrule(::typeof(*), A::AbstractSparseMatrix, B::DenseInputVecOrMat)
+function ChainRulesCore.rrule(
+    ::typeof(*), A::AbstractSparseMatrix{T}, B::DenseInputVecOrMat
+) where T 
     C = A * B
     function mul_pullback(C̄)
-        _, i_A, i_B, _, _ = (~imul!)(AD.GVar(C, C̄), AD.GVar(A), AD.GVar(B), AD.GVar(1.), AD.GVar(1.))
+        i_C, i_A, i_B, _, _ = (~imul!)(AD.GVar(C, unthunk(C̄)), AD.GVar(A), AD.GVar(B), AD.GVar(1.), AD.GVar(1.))
```

Comment:
We should instead add

```julia
@eval GVar(x::$T, y::AbstractThunk) = GVar(x, unthunk(y))
```

to https://github.com/GiggleLiu/NiLang.jl/blob/f24036c478ace467ae71f630e1db2698e38fd64d/src/autodiff/vars.jl#L72-L77

Reply:
Sorry... `i_C` is just used only in debug mode... I will delete it!

Reply:
@jieli-matrix 
Yeah, you need to add `unthunk`, this is how `ChainRules` people normally do:
https://github.com/JuliaDiff/ChainRules.jl/blob/26e46081e5210f4a81f41c9ab61ea466ca8dc006/src/rulesets/Base/evalpoly.jl#L4

You should not expect `GVar` to handle thunks for you due to the unclear gain and huge amount of work in the NiLang's infrastructure.